File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function(root) {
    if (!root) return 0;

    let maxiDepth = 0;
    const traverse = function(curr, depth) {
        maxiDepth = Math.max(maxiDepth, depth);

        curr.left && traverse(curr.left, depth + 1);
        curr.right && traverse(curr.right, depth + 1);
    }

    traverse(root, 1);

    return maxiDepth;
};

File: 0104-maximum-depth-of-binary-tree/0104-maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        
        maxiDepth = self.traverse(root, 0, 1)
        return maxiDepth
    
    def traverse(self, curr, maxiDepth, depth):
        maxiDepth = max(maxiDepth, depth)

        if curr and curr.left:
            maxiDepth = max(maxiDepth, self.traverse(curr.left, maxiDepth, depth + 1))
        
        if curr and curr.right:
            maxiDepth = max(maxiDepth, self.traverse(curr.right, maxiDepth, depth + 1))

        return maxiDepth

